main.cpp

```diff
@@ -70,29 +70,440 @@ write 3 UDTs below that EACH have:
+{
+    lightsCurrentState = !lightsCurrentState;
+    if (lightsCurrentState)
+    {
+        std::cout << "The lights are currently on."  << std::endl; 
+    } 
+    else
+    {
+        std::cout << "The lights are currently off."  << std::endl; 
+    }
+    
+    return lightsCurrentState;
+}
+
+
 /*
  new UDT 4:
  with 2 member functions
  */
 
+struct StudioComplex
+{
+    StudioComplex();
+    ~StudioComplex();
+
+    ControlRoom controlRoomA;
+    ControlRoom controlRoomB;
+    LiveRoom liveRoomA;
+    LiveRoom liveRoomB;
+
+    int bookSession(ControlRoom controlRoom, LiveRoom liveRoom, int hours);
+    int prepareInvoice(ControlRoom controlRoom, LiveRoom liveRoom, int hours, int rate);
+    void placeAdvertisement(std::string text); 
+};
+
+StudioComplex::StudioComplex() 
+{
+    std::cout << "Studio Complex being constructed." << std::endl; 
+}    
+
+StudioComplex::~StudioComplex() 
+{
+    std::cout << "Studio Complex being deconstructed." << std::endl; 
+    placeAdvertisement("ADVERTISEMENT: STUDIO FOR SALE!!!!");
+}    
+
+int StudioComplex::bookSession(ControlRoom controlRoom, LiveRoom liveRoom, int hours)
+{
+    std::cout <<  "Booking " << controlRoom.name << " with " << liveRoom.name << " for " << hours << " hours" << std::endl; 
+    return hours;
+}
+
+int StudioComplex::prepareInvoice(ControlRoom controlRoom, LiveRoom liveRoom, int hours, int rate)
+{
+    int cost = hours * rate;
+    std::cout << controlRoom.name << " with " << liveRoom.name << " fee is " << cost << " dollars" << std::endl; 
+    return cost;
+}
+
+void StudioComplex::placeAdvertisement(std::string text)
+{
+    std::cout << text << std::endl; 
+}
+
 /*
  new UDT 5:
  with 2 member functions
  */
 
+struct LargestFiveCities
+{
+    LargestFiveCities();
+    ~LargestFiveCities();
+
+    City cityA;
+    City cityB;
+    City cityC;
+    City cityD;
+    City cityE;
+
+    void setNames(std::string a, std::string b, std::string c, std::string d, std::string e);
+    int setPopulations(int a, int b, int c, int d, int e);
+};
+
+LargestFiveCities::LargestFiveCities() 
+{
+    std::cout << "LargestFiveCities being constructed." << std::endl; 
+}    
+
+LargestFiveCities::~LargestFiveCities() 
+{
+    std::cout << "LargestFiveCities being deconstructed." << std::endl; 
+}    
+
+void LargestFiveCities::setNames(std::string a, std::string b, std::string c, std::string d, std::string e)
+{
+    cityA.name = a;
+    cityB.name = b;
+    cityC.name = c;
+    cityD.name = d;
+    cityE.name = e;
+
+    std::cout << "The largest cities are " << a << ", " << b << ", " << c << ", " << d << ", and " << e << "." << std::endl; 
+}
+
+int LargestFiveCities::setPopulations(int a, int b, int c, int d, int e)
+{
+    cityA.population = a;
+    cityB.population = b;
+    cityC.population = c;
+    cityD.population = d;
+    cityE.population = e;
+
+    std::cout << "Combined population is " << a + b + c + d + e << std::endl; 
+    return a + b + c + d + e; 
```

Comment:
D.R.Y. 
cache this computed value locally
use it in the std::cout and return statements
- [ ] revise